File: core/src/com/epochgames/epoch/Epoch.java
package com.epochgames.epoch;

import com.badlogic.gdx.*;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.graphics.g2d.BitmapFont;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.utils.viewport.ExtendViewport;
import com.badlogic.gdx.utils.viewport.Viewport;
import com.epochgames.epoch.screens.InGame;
import com.epochgames.epoch.screens.MainMenu;
import com.epochgames.epoch.util.Assets;

public class Epoch extends Game {
    public static boolean debug = false;

	public Viewport viewport;
	public OrthographicCamera camera;

	public static GameManager gameManager;

	public BitmapFont font;
	public SpriteBatch batch;
	public SpriteBatch guiBatch;
	public boolean assetsLoaded = false;

	public Screen activeScreen;
	public InGame inGameScreen;
	public MainMenu mainMenuScreen;

	public Epoch(boolean debugMode) {
	    super();
	    debug = debugMode;
    }

	@Override
	public void create () {
		//Set the log level first thing so that we know what's going on
		Gdx.app.setLogLevel(debug ? Gdx.app.LOG_DEBUG : Gdx.app.LOG_INFO);

		//Loads Assets (Go figure)
		loadAssets();

		//Loading Assets
		while(!assetsLoaded) {
			//Use this to make a sick progress bar
			Gdx.app.log("Loading",  Assets.MANAGER.getProgress() * 100 + "% complete");

			//Updates assetsLoaded to true if all assets are loaded
			assetsLoaded = Assets.MANAGER.update();
		}

		//Create the GameManager
		gameManager = GameManager.getInstance();
		gameManager.setGame(this);

		//Setup the Camera, Viewport and Stage
		camera = new OrthographicCamera();
		viewport = new ExtendViewport(1920, 1080, camera);
		camera.setToOrtho(false, viewport.getScreenWidth(), viewport.getScreenHeight());

		//Create the Batches
		batch = new SpriteBatch();
		guiBatch = new SpriteBatch();
		font = new BitmapFont();

		//Set the screen to our starting screen and add all of them
		//TODO
		initializeScreens();
		gameManager.setGameState(GameManager.GameState.IN_GAME);
	}

	@Override
	public void render () {
		//Render the appropriate context by switching to the correct screen if we're not there already
		if(getScreen() != activeScreen) {
			setScreen(activeScreen);
		}

		//Calls the render() method of the appropriate screen
		super.render();
	}
	
	@Override
	public void dispose () {
		batch.dispose();
		guiBatch.dispose();
		font.dispose();
	}

	private void initializeScreens() {
		inGameScreen = new InGame(this);
		mainMenuScreen = new MainMenu(this);
	}

	private void loadAssets() {
		Assets.MANAGER.load(Assets.Spritesheets.SHIPS);
		Assets.MANAGER.load(Assets.TileMaps.OPEN_SPACE);
		Assets.MANAGER.load(Assets.TileMaps.OPEN_SPACE_TILE_SET);
		Assets.MANAGER.load(Assets.Textures.HEX_TILE);
	}

	public void setActiveScreen(Screen screen) {
		this.activeScreen = screen;
		setScreen(screen);
	}

	public void setActiveScreen(GameManager.GameState gameState) {
		Screen screen;
		switch (gameState) {
			case MAIN_MENU:
				screen = mainMenuScreen;
				break;
			case IN_GAME:
				screen = inGameScreen;
				break;
			default:
				Gdx.app.error("Invalid Screen", "An invalid game state was given");
				screen = null;
				break;
		}
		setActiveScreen(screen);
	}

	@Override
	public void resize(int width, int height) {
		viewport.update(width, height);
	}
}

File: core/src/com/ender/games/epoch/Epoch.kt
package com.ender.games.epoch

import com.badlogic.gdx.Application
import com.badlogic.gdx.Game
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.utils.viewport.ExtendViewport
import com.ender.games.epoch.screens.InGameScreen
import com.ender.games.epoch.screens.MainMenuScreen
import com.ender.games.epoch.util.*
import kotlin.properties.Delegates

class Epochkt(val screenWidth: Int,
              val screenHeight: Int,
              val debug: Boolean = false): Game() {

    val camera = OrthographicCamera()
    val viewport = ExtendViewport(192f / 4, 108f / 4, camera)

    lateinit var inGameScreen: InGameScreen
    lateinit var mainMenuScreen: MainMenuScreen

    override fun create() {
        GAME_MANAGER.game = this

        Gdx.app.logLevel = if(debug) Application.LOG_DEBUG else Application.LOG_INFO

        loadAssets()

        while(!ASSET_MANAGER.update()) {
            //Gdx.app.log("Loading", "${ASSET_MANAGER.progress * 100}% complete")
        }

        inGameScreen = InGameScreen(this)
        mainMenuScreen = MainMenuScreen(this)
        GAME_MANAGER.setGameState(GameState.MAIN_MENU)

        camera.setToOrtho(false, viewport.screenWidth.toFloat(), viewport.screenHeight.toFloat())

    }

    private fun loadAssets() {
        with(ASSET_MANAGER) {
            load(Spritesheets.SHIPS)
            load(Spritesheets.PLANETS)
            load(UI.SKIN)
            load(UI.ATLAS)
            load(Textures.LIGHT_AMMO_TEX)
            load(Textures.MED_AMMO_TEX)
            load(Textures.HEAVY_AMMO_TEX)
            load(Textures.TEMP_HEXACRON_TEST)
            load(Audio.MUSIC)
            load(Textures.HEX_UNIT)
            load(Textures.WALL_UNIT)
            load(Textures.CIRCLE_UNIT)
        }
    }

    fun setActiveScreen(gameState: GameState) {
        when (gameState) {
            // GameState.MAIN_MENU -> screen = mainMenuScreen
            GameState.IN_GAME -> setScreen(inGameScreen)
            GameState.MAIN_MENU -> setScreen(mainMenuScreen)
            else -> {
                Gdx.app.error("Invalid Screen", "An invalid game state was given")
                screen = null
            }
        }
    }

    override fun resize(width: Int, height: Int) = viewport.update(width, height)
}